diff --git a/AES_GCM_SIV_128/AES_GCM_SIV_128_Reference_Code/GCM_SIV_c.c b/AES_GCM_SIV_128/AES_GCM_SIV_128_Reference_Code/GCM_SIV_c.c
index 571a400..ee9ad82 100755
--- a/AES_GCM_SIV_128/AES_GCM_SIV_128_Reference_Code/GCM_SIV_c.c
+++ b/AES_GCM_SIV_128/AES_GCM_SIV_128_Reference_Code/GCM_SIV_c.c
@@ -298,7 +298,7 @@ void GCM_SIV_ENC_2_Keys(uint8_t* CT, uint8_t TAG[16], uint8_t K1[16], uint8_t N[
 	POLYVAL(LENBLK, (uint64_t*)HASH_KEY, 16, POLYV);
 	#ifdef XOR_WITH_NONCE
 	((uint64_t*)POLYV)[0] ^= ((uint64_t*)N)[0];
-	((uint64_t*)POLYV)[1] ^= ((uint64_t*)N)[1];
+	((uint64_t*)POLYV)[1] ^= ((uint32_t*)N)[2];
 	#endif
 	((uint8_t*)POLYV)[15] &= 127;
 	
@@ -377,7 +377,7 @@ int GCM_SIV_DEC_2_Keys(uint8_t* MSG, uint8_t TAG[16], uint8_t K1[16], uint8_t N[
 	new_TAG[1] = T[1];
 	#ifdef XOR_WITH_NONCE
 	((uint64_t*)new_TAG)[0] ^= ((uint64_t*)N)[0];
-	((uint64_t*)new_TAG)[1] ^= ((uint64_t*)N)[1];
+	((uint64_t*)new_TAG)[1] ^= ((uint32_t*)N)[2];
 	#endif
 	((uint8_t*)new_TAG)[15] &= 127;
 	AES_128_Encrypt((uint32_t*)new_TAG, (uint32_t*)new_TAG, (uint32_t*)KS); 
